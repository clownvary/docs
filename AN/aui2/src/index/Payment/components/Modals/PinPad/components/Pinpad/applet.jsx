import React from 'react';
import UIComponent from 'shared/components/UIComponent';
import './index.less';

const navigatorName = window.navigator.appName.toLowerCase();
const isIE = navigatorName.indexOf('internet explorer') !== -1;
const isNetScape = navigatorName.indexOf('netscape') !== -1;
/* eslint-disable */
export default class Applet extends UIComponent {
  render() {
    return (
      <div
        className="pinpad-transaction-fail"
        dangerouslySetInnerHTML={{ __html: this.renderApplet() }}
      />
    );
  }

  renderApplet() {
    const { applet_name, applets } = this.props.appletInfo;

    if (isIE || isNetScape) {
      if (isNetScape) {
        return `<applet
          ${applets}
          archive=${applet_name}.jar
          code=${applet_name}.class
          id=APDInterfaceApplet
          name=APDInterfaceApplet
          height=0
          width=0
          scriptable=true
          mayscript=mayscript
          pluginspage=http://java.com/en/download/index.jsp>
          ${this.renderParams()}
        </applet>`;
      }
      return `<object
        codebase={code_base}
        type=application/x-java-applet;version=1.5
        archive=${applet_name}.jar
        code=${applet_name}.class
        id=APDInterfaceApplet
        name=APDInterfaceApplet
        height=0
        width=0>
          <param name=mayscript value=true/>
          ${this.renderParams()}
        </object>`;
    }
    return '';
  }

  renderParams() {
    const {
      device_type_id,
      terminal_location_address,
      device_id,
      terminal_location_city,
      terminal_location_state,
      terminal_location_zip,
      card_acceptor_name,
      merchant_descriptor,
      merchant_user,
      merchant_password,
      ams_url,
      ams_port,
      ams_timeout,
      terminal_number,
      usb_port_name,
      pinpad_data_entry_timeout,
      pinpad_swipe_card_timeout,
      pinpad_pin_entry_timeout,
      write_to_log
    } = this.props.appletInfo;

    const ccScanWithoutEmv = device_type_id === 1 || device_type_id === 2;
    return `${ccScanWithoutEmv ? '<param name="separate_jvm" value="true"><param name="cache_option" value="No"><param name="classloader_cache" value="true">' : ''}
      <param name="dll_ver" value="1.0.0">
      <param name="dll_auto_update" value="Yes">
      <param name="AMSURL" value="${ams_url}">
      <param name="AMSPort" value="${ams_port}">
      <param name="AMSTimeout" value="${ams_timeout}">
      <param name="CardAcceptorName" value="${card_acceptor_name}">
      <param name="TerminalLocationAddress" value="${terminal_location_address}">
      <param name="TerminalLocationCity" value="${terminal_location_city}">
      <param name="TerminalLocationState" value="${terminal_location_state}">
      <param name="TerminalLocationZIP" value="${terminal_location_zip}">
      <param name="MerchantDescriptor" value="${merchant_descriptor}">
      <param name="MerchantUser" value="${merchant_user}">
      <param name="MerchantPassword" value="${merchant_password}">
      <param name="TerminalNumber" value="${terminal_number}">
      <param name="DeviceId" value="${device_id}">
      <param name="USBPortName" value="${usb_port_name}">
      <param name="PinpadDataEntryTimeout" value="${pinpad_data_entry_timeout}">
      <param name="PinpadPINEntryTimeout" value="${pinpad_pin_entry_timeout}">
      <param name="PinpadSwipeCardTimeout" value="${pinpad_swipe_card_timeout}">
      <param name="WriteToLog" value="${write_to_log}">`;
  }
}
/* eslint-enable */


/*
  the relationship between pinpad device type with applet name:
  {
    "1": "APDInterfaceApplet", //paymentech
    "2": "APDMonerisInterfaceApplet",
    "3": "APDEMVInterfaceApplet", //paymentech EMV
    "4": "APDEMVElavonInterfaceApplet"
  }
*/
