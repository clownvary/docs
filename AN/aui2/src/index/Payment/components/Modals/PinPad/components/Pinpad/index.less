@import '~shared/style/variables/index.less';
@imagePath: '~shared/images/iPP320.jpg';
@module-name: payment-pinpad;

.@{module-name} {
  p {
    padding: 0;
    margin: 0;
  }

  &-header {
    padding-bottom: 10px;
    overflow: hidden;
    border-bottom: 1px solid #d6d7da;

    span {
      &:first-child {
        float: left;
        font-weight: @semibold;
      }

      &:nth-child(2) {
        float: right;
        font-weight: 700;
      }
    }
  }

  &-body {
    padding-top: 30px;

    &.aaui-flexbox {
      align-items: flex-start;
      flex-wrap: nowrap;
    }

    &-img {
      width: 190px;
      height: 190px;
      font-size: 0;
      background-image: url(@imagePath);
      background-position: -20px 0;
      background-repeat: no-repeat;
      background-size: 100%;
    }

    &-right {
      height: 190px;
    }

    &-prompts {
      color: #748288;
    }
  }

  &-error-wrapper {
    &.aaui-flexbox {
      align-items: flex-end;
      flex-direction: column;
      flex-wrap: nowrap;
      justify-content: space-between;
      height: 100%;
    }

    .@{module-name}-retry {
      cursor: pointer;
    }

    .@{module-name}-error-alert {
      display: flex;
      width: 398px;

      .icon-close-circle,
      .icon-times-circle {
        padding-left: 10px;
        margin-top: 0;
        font-size: 16px;
        color: #db3811;
      }

      .error-content {
        display: inline-block;
        padding-right: 10px;
        padding-left: 10px;
        margin-top: -3px;
        vertical-align: top;
      }

      ul {
        padding: 0;
        margin: 0;
        list-style-type: none;
      }

      .@{module-name}-error-message {
        max-height: 65px;
        overflow: auto;
        font-size: 14px;
        line-height: 1.43;
      }
    }

    .alert-error {
      color: #db3811;
    }
  }

  &-error-action {
    line-height: 1.43;
    color: #748288;

    .payment-pinpad-retry {
      color: @color-blue;
    }

    .payment-pinpad-retry-separation {
      margin: 0 5px;
    }
  }
}
