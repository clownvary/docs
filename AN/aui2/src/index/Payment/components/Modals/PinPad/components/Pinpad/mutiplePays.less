@import '~shared/style/variables/index.less';
/* stylelint-disable */

.pinpad-mutiple-pays {
  .pinpad-wrapper {
    .payment-credit-debit-error {
      margin-bottom: 15px;

      .payment-pinpad-error-alert {
        width: 100%;
      }
    }
  }

  .payment-pinpad {
    padding: 15px;
    overflow: hidden;
    border-radius: 5px;

    & + .payment-pinpad {
      margin-top: 15px;
    }
  }

  .pinpad-active {
    height: 199px;
    border: solid 1px @color-blue;

    .payment-pinpad-body {
      padding-top: 15px;
    }

    .payment-pinpad-error-alert {
      width: 355px;
    }
  }

  .pinpad-collapse {
    height: 54px;
    overflow: hidden;
    cursor: pointer;
    border: solid 1px #ddd;

    &:hover {
      background-color: #eef9fd;
      border: solid 1px #cfdbe0;
    }

    &.pinpad-success,
    &.pinpad-fail {
      height: 74px;

      .payment-pinpad-header {
        span:first-child:after {
          display: inline-block;
          width: 14px;
          margin-left: 5px;
          font-family: 'aui_icons';
          vertical-align: bottom;
        }
      }
    }

    .payment-pinpad-header {
      border-bottom-width: 0;

      div {
        clear: both;
        height: 20px;
        font-size: 14px;
        line-height: 1.43;
        color: #748288;
      }
    }

    &.pinpad-fail {
      .payment-pinpad-header {
        span:first-child:after {
          color: #d93a20;
          content: '\F057';
        }
      }
    }

    &.pinpad-success {
      .payment-pinpad-header {
        span:first-child:after {
          color: #5cb85c;
          content: '\F058 ';
        }
      }
    }

    &.pinpad-disabled {
      color: #748288;
      background: #f3f3f3;
    }
  }

  .payment-pinpad-action-btns {
    div {
      margin-top: 30px;
      text-align: right;
    }
  }

  .pinpad-credit-debit-success-msg {
    margin: 30px 0 16px 0;
    text-align: center;
    color: #748288;
  }
}

.pinpad-modal .modal-body {
  position: relative;
}
