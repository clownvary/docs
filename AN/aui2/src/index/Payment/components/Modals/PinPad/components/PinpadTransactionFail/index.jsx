import React from 'react';
import { connect } from 'react-redux';
import Button from 'react-base-ui/lib/components/Button';
import { clearError } from 'shared/actions/Error';
import UIComponent from 'shared/components/UIComponent';

import {
  cancelCardPayment,
  cardLeaveBlance,
  showFailActionBtns
} from '../../actions/pinpadFail';

import {
  updateErrorMessage,
  gotoNextPage,
  clearPinpadPays
} from '../../actions/pinpad';

import { updateModalTitle, showModal } from '../../actions/pinpadModal';
import { titles } from '../../consts';
import './index.less';

export class PinpadTransactionFail extends UIComponent {
  render() {
    const { pinpadFail, payment } = this.props.pinpad;
    const { shown } = pinpadFail.toJS();
    const { apdPaymentInfo, isNewCard } = payment.toJS();

    let canCancelApdPayment = false;
    let canLeaveBalance = false;
    if (isNewCard) {
      canCancelApdPayment = true;
    } else {
      canCancelApdPayment = apdPaymentInfo.canCancelApdPayment;
      canLeaveBalance = apdPaymentInfo.canLeaveBalance;
    }

    return (
      <div className="pinpad-transaction-fail payment-pinpad-action-btns">
        {shown ?
          <div>
            {canCancelApdPayment ?
              <Button onClick={() => this.cancelPayment()}>Cancel</Button> : ''
            }

            {canLeaveBalance ?
              <Button onClick={() => this.leaveBalance()}>Leave Balance</Button> : ''
            }
          </div> : ''
        }
      </div>
    );
  }

  cancelPayment() {
    const that = this;
    const { pinpad: { payment, pinpadModal } } = that.props;
    const { apdInterfaceApplet } = pinpadModal.toJS();
    const {
      pays,
      batchID,
      receiptID,
      isNewCard
    } = payment.toJS();
    const paysLen = pays.length;

    that.props.clearError();

    that.props.cancelCardPayment({
      batchID,
      receiptID
    }, () => {
      apdInterfaceApplet && apdInterfaceApplet.APDInputEnd && apdInterfaceApplet.APDInputEnd();
      that.props.showModal(false);
      that.props.updateModalTitle(isNewCard ? titles.NEW_CARD_TITLE : titles.TRANSACTION_TITLE);
      that.props.showFailActionBtns(false);
      that.props.clearPinpadPays();
    }, (errorMsg) => {
      if (paysLen > 1) {
        that.props.updateErrorMessage({
          message: errorMsg,
          /*
            When has mutilple device payments and
            cancel the device payment will cause the server error
          */
          isServerErr: true
        });
      } else {
        that.props.updateErrorMessage({
          message: errorMsg
        });
      }
    });
  }

  leaveBalance() {
    const that = this;
    const payment = this.props.pinpad.payment.toJS();
    const {
      pays,
      apdPaymentInfo,
      batchID,
      receiptID
    } = payment;
    const {
      isPayerCashCustomer,
      hasImmediatePaymentDue
    } = apdPaymentInfo;
    const paysLen = pays.length;
    let errMsg = 'Payer is Drop-In customer. Cannot leave balance on account.';

    if (!isPayerCashCustomer && hasImmediatePaymentDue) {
      errMsg = 'Portion of balance is due immediately. Cannot leave balance on account.';
    }

    if (isPayerCashCustomer || hasImmediatePaymentDue) {
      if (paysLen > 1) {
        this.props.updateErrorMessage({
          message: errMsg,
          /*
            When has mutilple device payments and
            choose leave balance function excute fail will cause the server error
          */
          isServerErr: true
        });
      } else {
        this.props.updateErrorMessage({
          message: errMsg
        });
      }

      return false;
    }

    this.props.cardLeaveBlance({
      batchID,
      receiptID
    },
    this.props.gotoNextPage,
    (errorMsg) => {
      if (paysLen > 1) {
        that.props.updateErrorMessage({
          message: errorMsg,
          isServerErr: true
        });
      } else {
        that.props.updateErrorMessage({
          message: errorMsg
        });
      }
    });

    return false;
  }
}

export default connect(
  null,
  {
    cancelCardPayment,
    cardLeaveBlance,
    updateErrorMessage,
    updateModalTitle,
    showFailActionBtns,
    clearError,
    gotoNextPage,
    clearPinpadPays,
    showModal
  }
)(PinpadTransactionFail);
