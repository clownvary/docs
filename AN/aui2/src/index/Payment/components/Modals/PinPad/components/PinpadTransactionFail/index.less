.pinpad-transaction-fail {
  button {
    margin-left: 10px;
  }
}
