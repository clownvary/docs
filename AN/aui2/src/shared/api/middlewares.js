import PromiseMiddleware from './promiseMiddleware';
import RouteMiddleware from './routeMiddleware';
import ThunkMiddleware from './thunkMiddleware';
import API from './API';

export default [
  new ThunkMiddleware(),
  new RouteMiddleware(),
  new PromiseMiddleware(new API())
];
