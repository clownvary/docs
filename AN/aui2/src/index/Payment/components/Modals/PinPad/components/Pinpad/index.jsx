import Pinpad from './pinpad';
import SinglePay from './singlePay';
import MutiplePays from './mutiplePays';

import './index.less';

export default {
  Pinpad,
  SinglePay,
  MutiplePays
};
