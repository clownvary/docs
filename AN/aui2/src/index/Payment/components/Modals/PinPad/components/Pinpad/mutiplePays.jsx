import React from 'react';
import { connect } from 'react-redux';
import Alert from 'react-base-ui/lib/components/Alert';
import UIComponent from 'shared/components/UIComponent';
import Pinpad from './pinpad';
import PinpadTransactionFail from '../PinpadTransactionFail';
import { showFailActionBtns } from '../../actions/pinpadFail';
import './mutiplePays.less';
import {
  communicatingWithDevice,
  updateErrorMessage,
  updateCurrentPayIndex
} from '../../actions/pinpad';

function packagePays(paysArr, index, paymentErrMsg, communicating) {
  return paysArr.map((pay, key) => {
    const item = Object.assign({}, pay);

    let className = '';

    if (item.swipeSuccess) {
      item.headerHtml = item.successInfo;
      item.className = 'pinpad-success pinpad-collapse';
      return item;
    }

    if (index === key) {
      if (paymentErrMsg) {
        className += ' pinpad-collapse';

        /* istanbul ignore else */
        if (item.hasError) {
          className += ' pinpad-fail';
          item.className = className;
          item.headerHtml = 'No card information';
        }

        return item;
      }

      item.className = 'pinpad-active';
      return item;
    }

    className += ' pinpad-collapse';

    if (item.hasError) {
      className += ' pinpad-fail';
      item.headerHtml = 'No card information';
    }

    if (communicating) {
      className += ' pinpad-disabled';
    }

    item.className = className;
    return item;
  });
}

export class MutiplePays extends UIComponent {
  render() {
    const { pinpad } = this.props;
    const { payment } = pinpad;
    const { errorMessage, pays: paysOrg, currentPayIndex, communicating } = payment.toJS();
    const pays = packagePays(paysOrg, currentPayIndex, errorMessage, communicating);

    return (
      <div className="pinpad-mutiple-pays">
        <div className="pinpad-wrapper">
          {errorMessage ?
            <div className="payment-pinpad-error-wrapper aaui-flexbox payment-credit-debit-error">
              <Alert type="error" className="payment-pinpad-error-alert" noClose>
                <div className="error-content">
                  <div className="payment-pinpad-error-message">
                    <div>{errorMessage}</div>
                  </div>
                </div>
              </Alert>
            </div> : ''
          }
          {pays.map((item, key) => (
            <Pinpad
              key={`pay_${key}`}
              pinpad={pinpad}
              pay={item}
              onClick={() => this.onClick(key, item)}
            />
            ))}
        </div>
        <PinpadTransactionFail pinpad={pinpad} />
      </div>
    );
  }

  onClick(payIndex, payItem) {
    const { pinpad: { payment, pinpadModal } } = this.props;
    const { apdInterfaceApplet } = pinpadModal.toJS();
    const {
      pays,
      communicating,
      currentPayIndex,
      errorMessage,
      debitCardId
    } = payment.toJS();
    const clickPay = pays[payIndex];
    const {
      swipeSuccess,
      paymentTypeId
    } = clickPay;

    if (communicating) {
      return false;
    }

    if (!swipeSuccess) {
      if (currentPayIndex === payIndex) {
        if (payItem.className.indexOf('pinpad-collapse') > -1) {
          setTimeout(() => { // wait the fail action take work
            apdInterfaceApplet.APDInputStart(false, paymentTypeId === debitCardId);
          }, 100);
        }

        return false;
      }

      /* istanbul ignore else */
      if (errorMessage) {
        this.props.updateErrorMessage({
          message: '',
          isServerErr: true
        });
      }

      this.props.showFailActionBtns(false);
      this.props.communicatingWithDevice(true);
      this.props.updateCurrentPayIndex(payIndex);
    }

    return false;
  }
}

export default connect(
  null,
  {
    communicatingWithDevice,
    updateErrorMessage,
    showFailActionBtns,
    updateCurrentPayIndex
  }
)(MutiplePays);
