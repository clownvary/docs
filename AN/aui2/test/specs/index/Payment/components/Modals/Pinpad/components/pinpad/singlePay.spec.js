import React from 'react';
import toJson from 'enzyme-to-json';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import middlewares from 'shared/api/middlewares';
import SinglePay from 'index/Payment/components/Modals/PinPad/components/Pinpad/singlePay';
import Pinpad from 'index/Payment/components/Modals/PinPad/components/Pinpad/pinpad';
import PinpadTransactionFail from 'index/Payment/components/Modals/PinPad/components/PinpadTransactionFail';

jest.mock('index/Payment/components/Modals/PinPad/components/Pinpad/pinpad', () => 'Pinpad');
jest.mock('index/Payment/components/Modals/PinPad/components/PinpadTransactionFail', () => 'PinpadTransactionFail');

describe('index/Payment/components/Modals/PinPad/components/Pinpad/singlePay', () => {

  let store;

  beforeEach(() => {
    store = configureStore(middlewares)({});
  });

  const props = {
    pinpad: {},
    pay: {}
  };

  it('component should render correctly', () => {
    const component = mount(<SinglePay store={store} {...props} />);
    expect(toJson(component)).toMatchSnapshot();
  });
});
