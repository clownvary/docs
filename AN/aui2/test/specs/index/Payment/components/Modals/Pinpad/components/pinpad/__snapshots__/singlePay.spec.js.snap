// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index/Payment/components/Modals/PinPad/components/Pinpad/singlePay component should render correctly 1`] = `
<SinglePay
  pay={Object {}}
  pinpad={Object {}}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "subscribe": [Function],
    }
  }
>
  <Pinpad
    pay={Object {}}
    pinpad={Object {}}
  >
    <PinpadTransactionFail
      pinpad={Object {}}
    />
  </Pinpad>
</SinglePay>
`;
